File: server/src/main/java/dev/slimevr/config/TrackerConfig.java
package dev.slimevr.config;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import dev.slimevr.tracking.trackers.Tracker;
import io.github.axisangles.ktmath.Quaternion;


public class TrackerConfig {

	private String customName;
	private String designation;
	private boolean hide;
	private Quaternion adjustment;
	private Quaternion mountingOrientation;
	private Boolean allowDriftCompensation;


	public TrackerConfig() {
	}

	public TrackerConfig(Tracker tracker) {
		this.designation = tracker.getTrackerPosition()
			!= null ? tracker.getTrackerPosition().getDesignation() : null;
		this.customName = tracker.getCustomName();
		allowDriftCompensation = tracker.isImu();
	}

	static JsonNode toV2(JsonNode v1, JsonNodeFactory factory) {
		ObjectNode node = factory.objectNode();
		if (v1.has("customName"))
			node.set("customName", v1.get("customName"));
		if (v1.has("designation"))
			node.set("designation", v1.get("designation"));
		if (v1.has("hide"))
			node.set("hide", v1.get("hide"));
		if (v1.has("mountingRotation"))
			node.set("mountingRotation", v1.get("mountingRotation"));
		if (v1.has("adjustment"))
			node.set("adjustment", v1.get("adjustment"));
		return node;
	}

	public String getCustomName() {
		return customName;
	}

	public void setCustomName(String customName) {
		this.customName = customName;
	}

	public String getDesignation() {
		return designation;
	}

	public void setDesignation(String designation) {
		this.designation = designation;
	}

	public boolean isHide() {
		return hide;
	}

	public void setHide(boolean hide) {
		this.hide = hide;
	}

	public Quaternion getAdjustment() {
		return adjustment;
	}

	public void setAdjustment(Quaternion adjustment) {
		this.adjustment = adjustment;
	}

	public Quaternion getMountingOrientation() {
		return mountingOrientation;
	}

	public void setMountingOrientation(Quaternion mountingOrientation) {
		this.mountingOrientation = mountingOrientation;
	}

	public Boolean getAllowDriftCompensation() {
		return allowDriftCompensation;
	}

	public void setAllowDriftCompensation(Boolean allowDriftCompensation) {
		this.allowDriftCompensation = allowDriftCompensation;
	}
}

File: server/core/src/main/java/dev/slimevr/config/TrackerConfig.kt
package dev.slimevr.config

import dev.slimevr.VRServer
import dev.slimevr.config.serializers.QuaternionSerializer
import dev.slimevr.tracking.trackers.Tracker
import io.github.axisangles.ktmath.ObjectQuaternion
import kotlinx.serialization.Serializable

@Serializable
class TrackerConfig {
	var customName: String? = null
	var designation: String? = null

	@Serializable(with = QuaternionSerializer::class)
	var mountingOrientation: ObjectQuaternion? = null

	@Serializable(with = QuaternionSerializer::class)
	var mountingResetOrientation: ObjectQuaternion? = null

	/**
	 * Only checked if [ServerConfig.useMagnetometerOnAllTrackers] enabled
	 */
	var shouldHaveMagEnabled: Boolean? = null

	constructor(tracker: Tracker) {
		this.designation = if (tracker.trackerPosition != null) tracker.trackerPosition!!.designation else null
		this.customName = tracker.customName
		shouldHaveMagEnabled = tracker.isImu()
	}
}

// val Tracker.config: TrackerConfig
// 	get() = VRServer.instance.configManager.vrConfig.getTracker(this)
